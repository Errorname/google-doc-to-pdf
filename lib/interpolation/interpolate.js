"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dot_1 = require("./dot");
const interpolate = (doc, data, formatters) => {
  const placeholders = findPlaceholders(doc);
  return computeUpdates(placeholders, data, formatters);
};
const findPlaceholders = doc => {
  const placeholders = [];
  const processContent = content => {
    content.map(c => {
      if (c.paragraph) {
        c.paragraph.elements.map(e => {
          if (e.textRun) {
            const matches = e.textRun.content.match(/{{([^}]*)}}/gi) || [];
            matches.map(m => placeholders.push(m.slice(2, -2)));
          }
        });
      }
      if (c.table) {
        c.table.tableRows.map(r => {
          r.tableCells.map(c => {
            processContent(c.content);
          });
        });
      }
    });
  };

  if (doc.headers) {
    Object.keys(doc.headers).forEach(key => {
      processContent(doc.headers[key].content);
    });
  }

  if (doc.body) {
    processContent(doc.body.content);
  }

  return placeholders;
};
const availableFormatters = {
  lowercase: s => s.toLowerCase(),
  uppercase: s => s.toUpperCase()
};
const computeUpdates = (placeholders, data, formatters) => {
  formatters = Object.assign({}, availableFormatters, formatters);
  const replacements = placeholders.map(placeholder => {
    const computed = `${dot_1.default(data, placeholder, { formatters })}`;
    return [placeholder, computed];
  });
  return replacements.map(([placeholder, computed]) => ({
    replaceAllText: {
      replaceText: computed,
      containsText: {
        text: `{{${placeholder}}}`,
        matchCase: false
      }
    }
  }));
};
exports.default = interpolate;
