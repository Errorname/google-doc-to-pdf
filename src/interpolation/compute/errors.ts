export class UnknownFormatterError extends Error {}

export class UnvalidFormatterError extends Error {}

export class UndefinedVariableError extends Error {}
