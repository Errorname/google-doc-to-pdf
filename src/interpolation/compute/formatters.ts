import { Formatters } from '../types'

const defaultFormatters: Formatters = {
  lowercase: (s: string) => (s || '').toLowerCase(),
  uppercase: (s: string) => (s || '').toUpperCase(),
  capitalize: (s: string) =>
    s
      ? s
          .split(' ')
          .map(([head, ...tail]) => head.toUpperCase() + tail.join('').toLowerCase())
          .join(' ')
      : '',
  money: (s: string, locale: string, currencyISO: string, fractionDigits = 0) => {
    const options = currencyISO ? { style: 'currency', currency: currencyISO } : {}
    return Number(s || '').toLocaleString(locale || 'en', {
      ...options,
      minimumFractionDigits: fractionDigits
    })
  },
  image: (url: string, width: number, height: number) => ({
    url,
    width,
    height
  })
}

export default defaultFormatters
