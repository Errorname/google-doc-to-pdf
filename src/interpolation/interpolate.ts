import { GDoc, Request, Formatters, Placeholder, ContentPlaceholder } from './types'

import computeContent from './compute'

const findPlaceholders = (doc: GDoc): Placeholder[] => {
  const placeholders: Placeholder[] = []

  doc.body.content.map(c => {
    if (c.paragraph) {
      c.paragraph.elements.map(e => {
        if (e.textRun) {
          let textRun = e.textRun.content
          const matches = textRun.match(/{{([^}]*)}}/gi) || []
          for (let match of matches) {
            const start = e.startIndex + textRun.indexOf(match)
            textRun = textRun.replace(match, ''.padStart(match.length))
            placeholders.push({
              raw: match,
              position: {
                start,
                end: start + match.length
              }
            })
          }
        }
      })
    }
  })

  return placeholders
}

const analyzePlaceholders = (
  placeholders: Placeholder[],
  data: Record<string, any>,
  options?: {
    formatters?: Formatters
  }
): Placeholder[] => {
  return placeholders.map(placeholder => {
    const rawWithoutBrackets = placeholder.raw.slice(2, -2).trim()

    // CommentPlaceholder
    if (rawWithoutBrackets.startsWith('#')) {
      return {
        ...placeholder,
        type: 'comment',
        value: rawWithoutBrackets.slice(1).trim()
      }
    }

    if (['%', '@'].includes(rawWithoutBrackets[0])) {
      // Reserved for later usage
      return placeholder
    }

    // ContentPlaceholder
    return computeContent(placeholder, data, options)
  })
}

const buildUpdates = (placeholders: Placeholder[]): Request[] => {
  const updates: Request[] = []

  for (let placeholder of placeholders) {
    // ContentPlaceholder
    if (placeholder.type == 'content') {
      const contentPlaceholder = placeholder as ContentPlaceholder
      // Text
      if (contentPlaceholder.output === `${contentPlaceholder.output}`) {
        updates.push({
          replaceAllText: {
            replaceText: `${contentPlaceholder.output}`,
            containsText: {
              text: placeholder.raw,
              matchCase: true
            }
          }
        })
      }
    }

    // CommentPlaceholder
    else if (placeholder.type == 'comment') {
      updates.push({
        replaceAllText: {
          replaceText: '',
          containsText: {
            text: placeholder.raw,
            matchCase: true
          }
        }
      })
    }
  }

  return updates
}

const interpolate = (doc: GDoc, data: any, formatters: Formatters): Placeholder[] => {
  let placeholders = findPlaceholders(doc)
  return analyzePlaceholders(placeholders, data, { formatters })
}

export default interpolate
export { findPlaceholders, analyzePlaceholders, buildUpdates }
