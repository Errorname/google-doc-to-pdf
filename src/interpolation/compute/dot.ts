import { UndefinedVariableError } from './errors'

export default (path: string, data: any): any => {
  const iterative: string[] = []
  path.split('.').map(subPath => {
    const selector = subPath.match(/\[.*\]/)
    if (selector) {
      if (!subPath.endsWith(selector[0])) {
        throw new SyntaxError(`Couldn't parse ${path}`)
      }
      subPath = subPath.replace(selector[0], '')
      if (subPath) {
        iterative.push(subPath)
      }
      iterative.push(selector[0].slice(1, -1))
    } else {
      iterative.push(subPath)
    }
  })

  let prop = data
  for (let accessor of iterative) {
    if (prop[accessor] === undefined) {
      throw new UndefinedVariableError(`${accessor} is undefined.`)
    }
    prop = prop[accessor]
  }

  return prop
}
