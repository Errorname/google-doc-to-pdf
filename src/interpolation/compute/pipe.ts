import { Formatters } from '../types'
import defaultFormatters from './formatters'
import dot from './dot'
import { UnvalidFormatterError, UnknownFormatterError } from './errors'

export default (
  value: any,
  transformation: string,
  data: any,
  options?: { formatters?: Formatters; fallback?: string }
): any => {
  const formatters: Formatters = { ...defaultFormatters, ...(options && options.formatters) }
  const matches = transformation.match(/^(?<formatter>[^\(]+)(\((?<params>.+)?\))?/)

  if (!matches || !matches.groups || !matches.groups.formatter) {
    throw new UnvalidFormatterError(`${transformation} is not a valid formatter.`)
  }

  const formatterName = matches.groups.formatter.toLowerCase()
  const formatter = formatters[formatterName]

  if (!formatter) {
    throw new UnknownFormatterError(`${formatterName} is not defined.`)
  }

  let typedParams = []
  if (matches.groups.params) {
    typedParams = normalizeParams(matches.groups.params.split(','), data, options)
  }

  return formatter(value, ...typedParams)
}

const normalizeParams = function(
  untypedParams: string[],
  data: any,
  options?: { formatters?: Formatters; fallback?: string }
) {
  return untypedParams
    .map(e => e.trim())
    .map(e => {
      if (/^(\“.*\”|\".*\"|\‘.*\’|\'.*\')$/.test(e)) {
        // String
        return e.slice(1, -1)
      } else if (/^[-+]?([0-9]+|[0-9]+\.[0-9]*|[0-9]*\.[0-9]+)$/.test(e)) {
        // Number
        return Number(e)
      } else if (/^true|false$/.test(e)) {
        //Boolean
        return Boolean(e)
      } else {
        // Variable
        try {
          return dot(e, data)
        } catch (err) {
          return undefined
        }
      }
    })
}
