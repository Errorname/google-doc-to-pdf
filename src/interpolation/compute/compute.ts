import dot from './dot'
import pipe from './pipe'
import { Placeholder, ContentPlaceholder, Formatters } from '../types'

export default (
  placeholder: Placeholder,
  data: Object,
  options?: { formatters?: Formatters; fallback?: string }
): ContentPlaceholder => {
  const rawWithoutBrackets = placeholder.raw.slice(2, -2).trim()

  const [inputRaw, ...transformations] = rawWithoutBrackets.split('|').map(s => s.trim())

  const input = evaluateInput(inputRaw, data, options)

  const pipes = []
  let transformedValue = input.value
  for (let transformation of transformations) {
    const pipe = transformValue(transformedValue, transformation, data, options)
    pipes.push(pipe)
    transformedValue = pipe.output
  }

  const output = pipes.length == 0 ? input.value : pipes[pipes.length - 1].output

  return {
    ...placeholder,
    type: 'content',
    input,
    pipes,
    output
  } as ContentPlaceholder
}

const evaluateInput = (
  raw: string,
  data: Object,
  options?: { fallback?: string }
): { raw: string; error?: Error; value: any } => {
  try {
    return { raw, value: dot(raw, data) }
  } catch (error) {
    if ((options && options.fallback) === undefined) {
      return { raw, error, value: '' }
    } else {
      return { raw, error, value: options && options.fallback }
    }
  }
}

const transformValue = (
  value: any,
  transformation: string,
  data: Object,
  options?: { formatters?: Formatters; fallback?: string }
) => {
  try {
    return {
      raw: transformation,
      output: pipe(
        value,
        transformation,
        data,
        options
      )
    }
  } catch (error) {
    // Ignore unknown/invalid formatters
    return { raw: transformation, error, output: value }
  }
}
