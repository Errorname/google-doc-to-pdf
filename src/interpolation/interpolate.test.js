import interpolate, { findPlaceholders, analyzePlaceholders, buildUpdates } from './interpolate'
import { UndefinedVariableError, UnknownFormatterError } from './compute/errors'

describe('interpolate', () => {
  test('full document', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                { textRun: { content: 'Hello {{ user.name }}!' }, startIndex: 0, endIndex: 21 },
                {}
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{movie.title | uppercase}}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: '{{ movie.poster | image }}' },
                  startIndex: 124,
                  endIndex: 150
                }
              ]
            }
          }
        ]
      }
    }
    const result = interpolate(document, {
      user: { name: 'Thibaud' },
      movie: { title: 'Lost in Translation', rating: 5, poster: 'http://example.img' }
    })
    const updates = buildUpdates(result)
    expect(result).toMatchSnapshot()
    expect(updates).toMatchSnapshot()
  })

  test('full document — strict - Unknown formatter', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Hello {{ user.name | pretty }}!' },
                  startIndex: 0,
                  endIndex: 21
                },
                {}
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{movie.title | uppercase}}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: '{{ movie.poster | image }}' },
                  startIndex: 124,
                  endIndex: 150
                }
              ]
            }
          }
        ]
      }
    }
    expect(() => {
      const result = interpolate(
        document,
        {
          user: { name: 'Thibaud' },
          movie: { title: 'Lost in Translation', rating: 5, poster: 'http://example.img' }
        },
        {},
        true
      )
    }).toThrow(UnknownFormatterError)
  })

  test('full document — strict - Undefined variable', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Hello {{ user.surname }}!' },
                  startIndex: 0,
                  endIndex: 21
                },
                {}
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{ movie.title | uppercase }}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: '{{ movie.poster | image }}' },
                  startIndex: 124,
                  endIndex: 150
                }
              ]
            }
          }
        ]
      }
    }
    expect(() => {
      const result = interpolate(
        document,
        {
          user: { name: 'Thibaud' },
          movie: { title: 'Lost in Translation', rating: 5, poster: 'http://example.img' }
        },
        {},
        true
      )
    }).toThrow(UndefinedVariableError)
  })

  test('full document — strict - First error is thrown first', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Hello {{ user.surname }}!' },
                  startIndex: 0,
                  endIndex: 21
                },
                {}
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{ movie.title | pretty }}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: '{{ movie.poster | image }}' },
                  startIndex: 124,
                  endIndex: 150
                }
              ]
            }
          }
        ]
      }
    }
    expect(() => {
      const result = interpolate(
        document,
        {
          user: { name: 'Thibaud' },
          movie: { title: 'Lost in Translation', rating: 5, poster: 'http://example.img' }
        },
        {},
        true
      )
    }).toThrow(UndefinedVariableError)
  })
})

describe('findPlaceholders', () => {
  test('empty document', () => {
    const document = {
      body: {
        content: []
      }
    }
    const placeholders = findPlaceholders(document)

    expect(placeholders).toEqual([])
  })

  test('document with unknown content', () => {
    const document = {
      body: {
        content: [{}, {}]
      }
    }
    const placeholders = findPlaceholders(document)

    expect(placeholders).toEqual([])
  })

  test('single paragraph with no placeholders', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [{ textRun: { content: 'Hello world!' } }]
            }
          }
        ]
      }
    }
    const placeholders = findPlaceholders(document)

    expect(placeholders).toEqual([])
  })

  test('multiple paragraphs with multiple placeholders', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                { textRun: { content: 'Hello {{ user.name }}!' }, startIndex: 0, endIndex: 21 }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{movie.title | uppercase}}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          }
        ]
      }
    }
    const placeholders = findPlaceholders(document)

    expect(placeholders).toMatchSnapshot()
  })
})

describe('analyzePlaceholders', () => {
  test('no placeholders', () => {
    const placeholders = analyzePlaceholders([], {})

    expect(placeholders).toEqual([])
  })

  test('contentPlaceholder - text', () => {
    const placeholders = [
      {
        raw: '{{ user.name.first | lowercase }}',
        position: {
          start: 134,
          end: 166
        }
      }
    ]

    const contentPlaceholders = analyzePlaceholders(placeholders, {
      user: { name: { first: 'Thibaud' } }
    })

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('contentPlaceholder - image', () => {
    const placeholders = [
      {
        raw: '{{ catUrl | image }}',
        position: {
          start: 134,
          end: 166
        }
      }
    ]

    const contentPlaceholders = analyzePlaceholders(placeholders, {
      catUrl: 'http://example.cat'
    })

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('contentPlaceholder - image (width)', () => {
    const placeholders = [
      {
        raw: '{{ catUrl | image(150) }}',
        position: {
          start: 134,
          end: 166
        }
      }
    ]

    const contentPlaceholders = analyzePlaceholders(placeholders, {
      catUrl: 'http://example.cat'
    })

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('contentPlaceholder - image (with, height)', () => {
    const placeholders = [
      {
        raw: '{{ catUrl | image(150, 80) }}',
        position: {
          start: 134,
          end: 166
        }
      }
    ]

    const contentPlaceholders = analyzePlaceholders(placeholders, {
      catUrl: 'http://example.cat'
    })

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('commentPlaceholder', () => {
    const placeholders = [
      {
        raw: '{{ # This is a comment }}',
        position: {
          start: 134,
          end: 158
        }
      }
    ]

    const contentPlaceholders = analyzePlaceholders(placeholders, {})

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('unknownPlaceholder - reserved symbol', () => {
    const placeholders = [
      {
        raw: '{{ % reserved symbol }}',
        position: {
          start: 134,
          end: 158
        }
      }
    ]

    const contentPlaceholders = analyzePlaceholders(placeholders, {})

    expect(contentPlaceholders).toMatchSnapshot()
  })

  test('Unknown Formatter — strict', () => {
    const placeholders = [
      {
        raw: '{{ name | pretty }}',
        position: {
          start: 134,
          end: 153
        }
      }
    ]

    expect(() => {
      analyzePlaceholders(placeholders, { name: 'Antoine' }, { strict: true })
    }).toThrow(UnknownFormatterError)
  })

  test('Undefined variable — strict', () => {
    const placeholders = [
      {
        raw: '{{ surname | capitalize }}',
        position: {
          start: 134,
          end: 153
        }
      }
    ]

    expect(() => {
      analyzePlaceholders(placeholders, { name: 'Antoine' }, { strict: true })
    }).toThrow(UndefinedVariableError)
  })
})

describe('buildUpdates', () => {
  test('no placeholders', () => {
    const updates = buildUpdates([])

    expect(updates).toEqual([])
  })

  test('contentPlaceholder - text', () => {
    const contentPlaceholder = {
      raw: '{{ user.name.first }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'content',
      input: {
        raw: 'user.name.first',
        value: 'Thibaud'
      },
      pipes: [],
      output: 'Thibaud'
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - text (number)', () => {
    const contentPlaceholder = {
      raw: '{{ accounts[0].money }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'content',
      input: {
        raw: 'accounts[0].money',
        value: 1500
      },
      pipes: [],
      output: '1500'
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - image', () => {
    const contentPlaceholder = {
      raw: '{{ catUrl | image }}',
      position: {
        start: 42,
        end: 47
      },
      type: 'content',
      input: {
        raw: 'catUrl',
        value: 'http://example.cat'
      },
      pipes: [],
      output: {
        url: 'http://example.cat'
      }
    }

    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - image (width)', () => {
    const contentPlaceholder = {
      raw: '{{ catUrl | image(150) }}',
      position: {
        start: 42,
        end: 47
      },
      type: 'content',
      input: {
        raw: 'catUrl',
        value: 'http://example.cat'
      },
      pipes: [],
      output: {
        url: 'http://example.cat',
        width: 150
      }
    }

    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - image (width, height)', () => {
    const contentPlaceholder = {
      raw: '{{ catUrl | image(150, 80) }}',
      position: {
        start: 42,
        end: 47
      },
      type: 'content',
      input: {
        raw: 'catUrl',
        value: 'http://example.cat'
      },
      pipes: [],
      output: {
        url: 'http://example.cat',
        width: 150,
        height: 180
      }
    }

    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - unknown', () => {
    const contentPlaceholder = {
      raw: '{{ strange.type }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'content',
      input: {
        raw: 'strange.type',
        value: { test: 123 }
      },
      pipes: [],
      output: { test: 123 }
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('commentPlaceholder', () => {
    const contentPlaceholder = {
      raw: '{{ # This is a comment }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'comment'
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('unknownPlaceholder', () => {
    const contentPlaceholder = {
      raw: '{{ % Unknown placeholder }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'not_real_type'
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })
})
