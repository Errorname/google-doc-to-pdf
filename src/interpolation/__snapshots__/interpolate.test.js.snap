// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyzePlaceholders commentPlaceholder 1`] = `
Array [
  Object {
    "position": Object {
      "end": 158,
      "start": 134,
    },
    "raw": "{{ # This is a comment }}",
    "type": "comment",
    "value": "This is a comment",
  },
]
`;

exports[`analyzePlaceholders contentPlaceholder - text 1`] = `
Array [
  Object {
    "input": Object {
      "raw": "user.name.first",
      "value": "Thibaud",
    },
    "output": "thibaud",
    "pipes": Array [
      Object {
        "output": "thibaud",
        "raw": "lowercase",
      },
    ],
    "position": Object {
      "end": 166,
      "start": 134,
    },
    "raw": "{{ user.name.first | lowercase }}",
    "type": "content",
  },
]
`;

exports[`analyzePlaceholders unknownPlaceholder - reserved symbol 1`] = `
Array [
  Object {
    "position": Object {
      "end": 158,
      "start": 134,
    },
    "raw": "{{ % reserved symbol }}",
  },
]
`;

exports[`buildUpdates commentPlaceholder 1`] = `
Array [
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ # This is a comment }}",
      },
      "replaceText": "",
    },
  },
]
`;

exports[`buildUpdates contentPlaceholder - text 1`] = `
Array [
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ user.name.first }}",
      },
      "replaceText": "Thibaud",
    },
  },
]
`;

exports[`buildUpdates contentPlaceholder - unknown 1`] = `Array []`;

exports[`buildUpdates unknownPlaceholder 1`] = `Array []`;

exports[`findPlaceholders multiple paragraphs with multiple placeholders 1`] = `
Array [
  Object {
    "position": Object {
      "end": 21,
      "start": 6,
    },
    "raw": "{{ user.name }}",
  },
  Object {
    "position": Object {
      "end": 64,
      "start": 37,
    },
    "raw": "{{movie.title | uppercase}}",
  },
  Object {
    "position": Object {
      "end": NaN,
      "start": NaN,
    },
    "raw": "{{ movie.rating }}",
  },
]
`;

exports[`interpolate full document 1`] = `
Array [
  Object {
    "input": Object {
      "raw": "user.name",
      "value": "Thibaud",
    },
    "output": "Thibaud",
    "pipes": Array [],
    "position": Object {
      "end": 21,
      "start": 6,
    },
    "raw": "{{ user.name }}",
    "type": "content",
  },
  Object {
    "input": Object {
      "raw": "movie.title",
      "value": "Lost in Translation",
    },
    "output": "LOST IN TRANSLATION",
    "pipes": Array [
      Object {
        "output": "LOST IN TRANSLATION",
        "raw": "uppercase",
      },
    ],
    "position": Object {
      "end": 64,
      "start": 37,
    },
    "raw": "{{movie.title | uppercase}}",
    "type": "content",
  },
  Object {
    "input": Object {
      "raw": "movie.rating",
      "value": 5,
    },
    "output": 5,
    "pipes": Array [],
    "position": Object {
      "end": NaN,
      "start": NaN,
    },
    "raw": "{{ movie.rating }}",
    "type": "content",
  },
]
`;
